include/beast/http/file_body_linux.hpp

```diff
@@ -0,0 +1,256 @@
+//
+// Copyright (c) 2013-2017 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_HTTP_FILE_BODY_LINUX_HPP
+#define BEAST_HTTP_FILE_BODY_LINUX_HPP
+
+#ifdef __linux__
+# define BEAST_NO_LINUX_FILE 0
+#else
+# define BEAST_NO_LINUX_FILE 1
+#endif
+
+#if ! BEAST_NO_LINUX_FILE
+
+#include <beast/core/async_result.hpp>
+#include <beast/core/error.hpp>
+#include <beast/core/file_base.hpp>
+#include <beast/http/message.hpp>
+#include <beast/http/serializer.hpp>
+#include <boost/asio/buffer.hpp>
+#include <boost/asio/basic_stream_socket.hpp>
+#include <boost/filesystem.hpp>
+#include <csignal>
+#include <sys/sendfile.h>
+#include <unistd.h>
+
+namespace beast {
+namespace http {
+
+struct file_body_linux
+{
+    class reader;
+    class writer;
+
+    class value_type
+    {
+        int file_{-1};
+        std::uint64_t size_{0};
+
+        friend class reader;
+        friend class writer;
+
+        template<int Signal>
+        static void signal_init()
+        {
+            static auto x{std::signal(Signal, SIG_IGN)};
+            (void)x;
+        }
+
+    public:
+        value_type(const value_type&) = delete;
+        value_type& operator=(const value_type&) = delete;
+
+        /** Destructor.
+
+            If the file handle is open, it is closed first.
+        */
+        ~value_type()
+        {
+            if(file_ != -1)
+                ::close(file_);
+        }
+
+        /// Default constructor
+        value_type() = default;
+
+        /// Move constructor
+        value_type(value_type&& other) noexcept
+            : file_(other.file_)
+            , size_(other.size_)
+        {
+            other.file_ = -1;
+            other.size_ = 0;
+        }
+
+        /// Move assignment
+        value_type& operator=(value_type&& other) noexcept
+        {
+            value_type tmp(std::move(other));
+            std::swap(file_, tmp.file_);
+            std::swap(size_, tmp.size_);
```

Comment:
I think that the move assignment should specify that the moved-from object is left as-if default-constructed. You like keeping objects around (e.g. http_worker) and I imagine that the moved-from object could be left with an open file unintentionally. There's precedent for this, I believe that's how `basic_io_object` works (correct me if wrong). After a move you get a clean socket ready for reuse.